### ISSUE TITLE
Server-side prerendering via reconciler breaks with subcomponent query join

### ISSUE THREAD

**Issue by levitanong** [omcljs/om#782]
- I'm prerendering on a boot task, but it should be pretty much the same environment as a server. (it's just clj, after all) 
- I'm using `om/render-to-str` on a the return value of `om/add-root`.
  - The problem does not appear when i call `om/render-to-str` on `((om/factory Component))`.
- If I remove the query join, the build proceeds. 
- The problem only exists on subcomponent query joins. Root components don't seem to have this problem.

Below is the stack trace (note that boot_prerender.clj at the bottom of the stack trace is my boot task):

```
java.lang.IllegalArgumentException: contains? not supported on type: clojure.lang.Keyword
                                                ...
                clojure.core/contains?/invokeStatic            core.clj: 1489
                             clojure.core/contains?            core.clj: 1481
         om.next.impl.parser$path_meta.invokeStatic         parser.cljc:  211
               om.next.impl.parser$path_meta.invoke         parser.cljc:  167
      om.next.impl.parser$path_meta$fn__5984.invoke         parser.cljc:  186
                     clojure.core/map-indexed/fn/fn            core.clj: 7170
                                                ...
                clojure.core/transduce/invokeStatic            core.clj: 6758
                     clojure.core/into/invokeStatic            core.clj: 6772
                                  clojure.core/into            core.clj: 6762
         om.next.impl.parser$path_meta.invokeStatic         parser.cljc:  183
               om.next.impl.parser$path_meta.invoke         parser.cljc:  167
      om.next.impl.parser$path_meta$fn__5984.invoke         parser.cljc:  186
                     clojure.core/map-indexed/fn/fn            core.clj: 7170
                                                ...
                clojure.core/transduce/invokeStatic            core.clj: 6758
                     clojure.core/into/invokeStatic            core.clj: 6772
                                  clojure.core/into            core.clj: 6762
         om.next.impl.parser$path_meta.invokeStatic         parser.cljc:  183
               om.next.impl.parser$path_meta.invoke         parser.cljc:  167
         om.next.impl.parser$path_meta.invokeStatic         parser.cljc:  213
               om.next.impl.parser$path_meta.invoke         parser.cljc:  167
         om.next.impl.parser$path_meta.invokeStatic         parser.cljc:  173
               om.next.impl.parser$path_meta.invoke         parser.cljc:  167
       om.next.impl.parser$parser$self__6004.invoke         parser.cljc:  294
       om.next.impl.parser$parser$self__6004.invoke         parser.cljc:  246
             om.next.Reconciler$parsef__7744.invoke           next.cljc: 2437
                  om.next.Reconciler.add_root_BANG_           next.cljc: 2458
                om.next$add_root_BANG_.invokeStatic           next.cljc: 1540
                      om.next$add_root_BANG_.invoke           next.cljc: 1527
                om.next$add_root_BANG_.invokeStatic           next.cljc: 1534
                      om.next$add_root_BANG_.invoke           next.cljc: 1527
  bahay.boot-prerender/render-to-file!/invokeStatic  boot_prerender.clj:   27
               bahay.boot-prerender/render-to-file!  boot_prerender.clj:   23
bahay.boot-prerender/eval9772/fn/middleware/handler  boot_prerender.clj:   47
                                boot.core/run-tasks            core.clj:  938
                                  boot.core/boot/fn            core.clj:  948
                clojure.core/binding-conveyor-fn/fn            core.clj: 2019
```

**Comment by levitanong** [comment]
@anmonteiro Decided to post here instead of Slack for context.

```clj
;; data
(def data 
  {:people 
    [{:person/id 1
      :person/name "Dude McDudeman"
      :person/roles 
        [{:role/id :cof :role/name "Co-Founder"}
         {:role/id :dev :role/name "Developer"}]}]})

;; components and queries
(defui Role
  static om/Ident
  (ident [this {:keys [role/id]}]
    [:role/by-id id])
  static om/IQuery
  (query [this]
    [:role/id :role/name]))

(defui Person
  static om/Ident
  (ident [this {:keys [person/id]}]
    [:person/by-id id])
  static om/IQuery
  (query [this]
    [:person/id :person/name
     {:roles (om/get-query Role)}]))

(defui People
  static om/IQuery
  (query [this]
    [{:people (om/get-query Person)}])
  Object
  (render [this]
    (dom/div nil
      "Test")))
```

**Comment by anmonteiro** [comment]
Can't repro with the given data + queries

**Comment by levitanong** [comment]
@anmonteiro Have you tried this on a boot repl?

For convenience, below is some code you can easily copy-paste into a `boot repl`. This should be enough to bring up the error.

```clj

(require '[om.next :as om :refer [defui]])
(require '[om.dom :as dom])

(def init-state
  {:people
   [{:person/id 1
     :person/name "Dude McDudeman"
     :person/roles
     [{:role/id :cof :role/name "Co-Founder"}
      {:role/id :dev :role/name "Developer"}]}]})

(defui Role
  static om/Ident
  (ident [this {:keys [role/id]}]
    [:role/by-id id])
  static om/IQuery
  (query [this]
    [:role/id :role/name]))

(defui Person
  static om/Ident
  (ident [this {:keys [person/id]}]
    [:person/by-id id])
  static om/IQuery
  (query [this]
    [:person/id :person/name
     {:roles (om/get-query Role)}]))

(defui People
  static om/IQuery
  (query [this]
    [{:people (om/get-query Person)}])
  Object
  (render [this]
    (dom/div nil
      "Test")))

(def parser (om/parser {:read (fn [env key _] {:value (get @(:state env) key)})}))

(def reconciler (om/reconciler {:state init-state :parser parser}))

(def c (om/add-root! reconciler People nil))
```

**Comment by levitanong** [comment]
The problem disappears when the read function of the parser is
```clj
{:value (om/db->tree (get-query Person) (get st key) st}
```

In cljs, even if i had just used `{:value (get st key)}` it would still work, and just return the ident. It appears that in clj, the parser is a little pickier.

**Comment by anmonteiro** [comment]
fixed 10c325907d90a4ade3318a8e9480fc696c75c205

**Comment by levitanong** [comment]
Hurrah! Thanks! :D
